### ISSUE TITLE
ZFS Crypto support

### ISSUE THREAD

**Issue by lundman** [openzfsonosx/zfs#466]
Some work has been done on this over in ZOL camp, I have created a mirrored issue here, so we can have any OSX internal discussion regarding the porting and other information

The current branch of the code is under https://github.com/openzfsonosx/zfs/compare/wip_crypt3

```
# ./cmd.sh zpool create -f DOOM ~/diskpool.img
# ./cmd.sh zfs create -o encryption=aes-256-ccm -o keysource=raw,prompt DOOM/ccm
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
# cp /usr/share/dict/words /Volumes/DOOM/ccm/THIS.IS.ENCRYPTED.IN.THEORY
# head /Volumes/DOOM/ccm/THIS.IS.ENCRYPTED.IN.THEORY
aardvark
# strings -a ~/diskpool.img |grep -i aardvark
#
```

**Comment by lundman** [comment]
The current branch name is `zol_crypto`

Let's put the OSX porting problems in here instead of cluttering the ZOL issue at https://github.com/zfsonlinux/zfs/pull/4329

**Comment by lundman** [comment]
import -l should import and ask for keys

```
# ./cmd.sh zpool create -f DOOM disk1
# ./cmd.sh zfs create -o encryption=aes-256-ccm -o keysource=passphrase,prompt DOOM/ccm
# mkdir /Volumes/DOOM/ccm/ADVAARK.ADVAARK.ADVAARK.ADVAARK.ADVAARK.ADVAARK
# ./cmd.sh zpool export -a
# ./cmd.sh zpool import -l DOOM
cannot mount 'DOOM/ccm': Insufficient privileges
```

**Comment by lundman** [comment]
Loading keys `zfs key -l` and `zfs mount` should work:
```
# ./cmd.sh zpool import DOOM
# ./cmd.sh zfs key -l DOOM/ccm
Enter passphrase for 'DOOM/ccm':
# ./cmd.sh zfs mount -a
# ls -al /Volumes/DOOM/ccm
#
(Notice no "." nor "..")
```

**Comment by lundman** [comment]
Testing the ciphers, we have some inconsistency.

Input data:
```
0x00 0x01 0x02 0x03 0x04 0x05 0x06 0x07
0x08 0x09 0x0a 0x0b 0x0c 0x0d 0x0e 0x0f
```

encrypted EXPECTED output:
```
 0x03 0x54 0x08 0xa4 0x52 0xb8 0xde 0xc4
 0x96 0x30 0x13 0xa8 0x5e 0xc4 0x32 0xca
```

encrypted OBSERVED output:
```
0x03 0x54 0x08 0xa4 0x52 0xb8 0xde 0xc4
0x68 0xfa 0xce 0x12 0xa0 0x0e 0xef 0x70
```
decrypted output:
```
0x00 0x01 0x02 0x03 0x04 0x05 0x06 0x07
0xfe 0xca 0xdd 0xba 0xfe 0xca 0xdd 0xba
```

Most likely the UIO porting changes has a bug in them producing close, but incorrect data after the first round.

**Comment by lundman** [comment]
Corrected the uio port, and encryption comes out as:

```
 0x03 0x54 0x08 0xa4 0x52 0xb8 0xde 0xc4
 0x68 0xfa 0xce 0x12 0xa0 0x0e 0xef 0x70
 0x3d 0x79 0x8f 0xda 0xdf 0x80 0xd4 0x1f
```

and the MAC comes out correctly as well

```
 MAC output: 0xc4 0x3e 0x15 0x88 0xab 0x48 0x65 0x0d 0x7b 0x3b 0x2c 0x10 0x57 0x01 0x68 0x11
```

However, decryption is still wrong

```
 0x00 0x01 0x02 0x03 0x04 0x05 0x06 0x07
 0xfe 0xca 0xdd 0xba 0xfe 0xca 0xdd 0xba
```

What I am unsure about is why decryption does not call `crypto_init_ptrs` and `crypto_get_ptrs` - you'd think they would need to call the same.

**Comment by lundman** [comment]
Branch `zol_crypto` is now complete and awaits testing.

**Comment by lundman** [comment]
First panic from test run `zfs_create_014_pos`

```
panic(cpu 0 caller 0xffffff7f9e2b9200): "kernel heap corruption detected"@spl-kmem.c:880
Backtrace (CPU 0), Frame : Return Address
0xffffff90bb4d37f0 : 0xffffff801c6df792 mach_kernel : _panic + 0xe2
0xffffff90bb4d3870 : 0xffffff7f9e2b9200 net.lundman.spl : _kmem_error + 0x6b0
0xffffff90bb4d3970 : 0xffffff7f9e2b5a75 net.lundman.spl : _zfs_kmem_free + 0x1b5
0xffffff90bb4d39d0 : 0xffffff7f9e42c5ca net.lundman.zfs : _zio_crypt_key_destroy + 0xda
0xffffff90bb4d39f0 : 0xffffff7f9e364f19 net.lundman.zfs : _dsl_keychain_entry_free + 0x19
0xffffff90bb4d3a10 : 0xffffff7f9e362e7a net.lundman.zfs : _dsl_keychain_free + 0x4a
0xffffff90bb4d3a30 : 0xffffff7f9e362e0e net.lundman.zfs : _spa_keystore_keychain_rele + 0x6e
0xffffff90bb4d3a60 : 0xffffff7f9e363cf7 net.lundman.zfs : _spa_keystore_remove_keychain_record + 0xc7
0xffffff90bb4d3ad0 : 0xffffff7f9e34c5f5 net.lundman.zfs : _dsl_dataset_disown + 0x65
0xffffff90bb4d3b00 : 0xffffff7f9e32d6f0 net.lundman.zfs : _dmu_objset_disown + 0x20
0xffffff90bb4d3b20 : 0xffffff7f9e3f476f net.lundman.zfs : _zfs_vfs_unmount + 0x68f
0xffffff90bb4d3ce0 : 0xffffff801c907c5e mach_kernel : _dounmount + 0x3ae
```

**Comment by lundman** [comment]
```
Results Summary
FAIL	 135
SKIP	  30
KILLED	   3
PASS	 565

Running Time:	03:51:21
Percent passed:	77.1%
```

**Comment by lundman** [comment]
```
# ./cmd.sh zpool create -f -O encryption=aes-256-ccm DOOM ~/poolfile.img
```
```
0xffffff8c957db520 : 0xffffff8018fb54b3 mach_kernel : trap_from_kernel + 0x26
0xffffff8c957db540 : 0xffffff7f9ab16f86 net.lundman.zfs : _dsl_dataset_create_sync_dd + 0x7f6
0xffffff8c957db7d0 : 0xffffff7f9ab304c1 net.lundman.zfs : _dsl_pool_create + 0x3d1
0xffffff8c957db880 : 0xffffff7f9ab57b87 net.lundman.zfs : _spa_create + 0x7a7
0xffffff8c957db9e0 : 0xffffff7f9aba9cea net.lundman.zfs : _zfs_ioc_pool_create + 0x29a
0xffffff8c957dba70 : 0xffffff7f9abb0222 net.lundman.zfs : _zfsdev_ioctl + 0x632
0xffffff8c957dbb50 : 0xffffff8019105fc2 mach_kernel : _spec_ioctl + 0x112
0xffffff8c957dbb90 : 0xffffff80190f3e9e mach_kernel : _VNOP_IOCTL + 0x16e
```

**Comment by lundman** [comment]
@tcaputi Can I confirm that zpool create with encryption should work? I suspect it panics as it did not ask for passphrase and carried on with empty dcp.

**Comment by tcaputi** [comment]
It seems I have a bug in the error handling for `zpool create`. `zpool create` works if you specify a keysource, but crashes if you don't provide one. I will fix this bug when I get in to work.

**Comment by tcaputi** [comment]
So there seems to have been 2 issues here. First is that my userspace code was not performing the check correctly when using `zpool create` instead of `zfs create`. The second (and more concerning) issue is that the kernel wasn't calling the check function. I looked at the code a little bit (in ZoL) and, from what I can tell, the `zpool create` handling doesn't do ANY validation of the userspace input (for ANY property). This is outside of the scope of the ZFS encryption patch, but should probably be submitted as a bug. 

Anyway, the fixed code was pushed and rebased onto the ZoL PR. Here is the diff (for your porting purposes)

```
diff --git a/lib/libzfs/libzfs_crypto.c b/lib/libzfs/libzfs_crypto.c
index 7870331..b68dc32 100644
--- a/lib/libzfs/libzfs_crypto.c
+++ b/lib/libzfs/libzfs_crypto.c
@@ -737,13 +737,15 @@ zfs_crypto_create(libzfs_handle_t *hdl, char *parent_name, nvlist_t *props,
                        goto error;
                }
        } else {
-               if (!nvlist_exists(pool_props, "feature@encryption")) {
+               if (!nvlist_exists(pool_props, "feature@encryption") &&
+                   local_crypt) {
                        ret = EINVAL;
                        zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
                            "Encryption feature not enabled."));
+                       goto error;
                }
 
-               pcrypt = crypt;
+               pcrypt = ZIO_CRYPT_OFF;
        }
 
        /* Check for encryption being explicitly truned off */
diff --git a/lib/libzfs/libzfs_pool.c b/lib/libzfs/libzfs_pool.c
index 4c8f71d..e889e10 100644
--- a/lib/libzfs/libzfs_pool.c
+++ b/lib/libzfs/libzfs_pool.c
@@ -1224,6 +1224,7 @@ zpool_create(libzfs_handle_t *hdl, const char *pool, nvlist_t *nvroot,
                }
                if (zfs_crypto_create(hdl, NULL, zc_fsprops, props,
                    &hidden_args) != 0) {
+                       zfs_error(hdl, EZFS_CRYPTOFAILED, msg);
                        goto create_failed;
                }
                if (nvlist_add_nvlist(zc_props,

```

**Comment by lundman** [comment]
Awesome thanks, appreciate the patch too, since diffing between two force pushes is tricky

**Comment by lundman** [comment]
@tcaputi One final concern, could also be a porting issue, although, that port of the code went in without much massaging.

Creating some test data, make a few files containing just
```
# ls -lR testdata/
total 192096
drwxr-xr-x  3 root  admin       102 May 27 09:35 GRAVITATIONAL
-rw-r--r--  1 root  admin  32783104 May 27 09:34 fileA
-rw-r--r--  1 root  admin  32783104 May 27 09:35 fileB
-rw-r--r--  1 root  admin  32783104 May 27 09:35 fileC

testdata//GRAVITATIONAL:
total 64032
-rw-r--r--  1 root  admin  32783104 May 27 09:35 fileD

# head testdata/fileA
AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK
AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK

# grep -r AARDVARK testdata/| wc -l
 1659904
```
That's a lot of lines.



```
$ mkfile 1G poolfile.img    # make new empty image
$ strings -a poolfile.img    # confirm it is empty

# create encrypted pool
$ ./cmd.sh zpool create -O encryption=aes-256-ccm -O keysource=passphrase,prompt -f DOOM ~/poolfile.img

# Copy text files to it
$ rsync -ar testdata/ /Volumes/DOOM/

# Export pool
$ ./cmd.sh zpool export -a

# Check what we can read

$ strings -a ~/poolfile.img |grep -i AARDVARK
AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK
AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK..AARDVARK

$ strings -a ~/poolfile.img |grep -i AARDVARK | wc -l
     384
```

There are definitely some data on there that is plain text.

**Comment by tcaputi** [comment]
Ive run this exact test several times myself (although I like 'hiimtom' instead of 'AARDVARK') and have never seen this come up. I will run the test again tomorrow to see if I can reproduce this in ZoL.

**Comment by lundman** [comment]
Yes, I should probably also see if I can reproduce it with ZOL. I'll keep poking at it and see if I can find something more out.

**Comment by lundman** [comment]
Ok, checking vdev_file start_io, and pagein, and pageout, the first hit of the aardvark is

```
vdev_file writing aardvark! 0xffffff8c8b29b77e

(lldb) p *zio
(zio_t) $91 = {
  io_bookmark = (zb_objset = 21, zb_object = 0, zb_level = -2, zb_blkid = 11)
  io_prop = {
    zp_checksum = ZIO_CHECKSUM_INHERIT
    zp_compress = ZIO_COMPRESS_INHERIT
    zp_type = DMU_OT_NONE
    zp_level = '\0'
    zp_copies = '\0'
    zp_dedup = 0
    zp_dedup_verify = 0
    zp_nopwrite = 0
    zp_encrypt = 0
  }
```

which seems to suggest it has decided to not encrypt it, so it is as expected. I guess a better question is then why it has decided to go with plaintext.

**Comment by tcaputi** [comment]
That is a ZIL block (by the zb_level). ZIL blocks are not do not have their `io_prop` set via `dmu_write_policy()` like most other blocks. Instead, these blocks go through `zio_alloc_zil()`, which just allocates a block so that data can be written into it later via `zio_rewrite()`. This case should be handled by the following code in zio.c's `zio_write_bp_init()`:

```
		/*
		 * encrypted ZIL blocks still need to have their encryption
		 * parameters setup, even though the block is pre-allocated
		 */
		if (BP_GET_TYPE(bp) == DMU_OT_INTENT_LOG &&
		    BP_IS_ENCRYPTED(bp)) {
			encrypt = B_TRUE;
		} else {
			return (ZIO_PIPELINE_CONTINUE);
		}
```

I can look at this more tomorrow, but a good test would be to see if there is a time when `zb->io_bookmark.zb_objset` is an encrypted dataset and that `else` statement is entered instead of the `if`.

**Comment by lundman** [comment]
Yes, I'm starting to wonder if it is ZIL.

Checking the `if` - it gets 11 times before we have the bad write, but the `else` is never hit. 

I will check if the `zio_crypt_init_uios_zil()` call maybe sets up rogue buffers. That call was somewhat complicated to OSXify in zio_crypto.c.

**Comment by tcaputi** [comment]
> I will check if the zio_crypt_init_uios_zil() call maybe sets up rogue buffers. That call was somewhat complicated to OSXify in zio_crypto.c.

That very well could be the problem. Check that and let me know if you find any issues (I would look over it myslef but [the commit you linked](https://github.com/openzfsonosx/zfs/blob/wip_crypt3/module/zfs/zio_crypt.c#L533) seems to have an old version of the code (before that function came into existence).

**Comment by lundman** [comment]
OK, turns out past-lundman left some desperate debug work in there, ie

```
	if (ot == DMU_OT_INTENT_LOG) {
		return ZIO_NO_ENCRYPTION_NEEDED;
		ret = zio_crypt_init_uios_zil(encrypt, plainbuf, cipherbuf,
			datalen, puio, cuio, enc_len);
```

siiggh..

I will now confirm there is no plaintext after removing that return, thanks for the assist!

**Comment by lundman** [comment]
Ok, so first problem was that `return` because it was causing a panic when ZIL encryption was on. And the reason for the panic was the "fast forward" of UIOs to offset had:
```
		for (vec_idx = 0;
			 uio_getiov(uiop, vec_idx, NULL, &iov_len) &&
			 vec_idx < uio_iovcnt(uiop) && offset >= iov_len;
			 offset -= iov_len, vec_idx++)
			;
```
instead of 
```
		for (vec_idx = 0;
			 !uio_getiov(uiop, vec_idx, NULL, &iov_len) &&
			 vec_idx < uio_iovcnt(uiop) && offset >= iov_len;
			 offset -= iov_len, vec_idx++)
			;
```

Ie, the **!** for `uio_getiov()` would always stop the for at the first iov. This works with normal encryption (only 1 buffer) but with ZIL it has many, and would compute a negative length, and attempt a bcopy of ludicrous size. 


I have now confirmed there is no plaintext leakage.

**Comment by noberasco** [comment]
keep on the good work guys!

I'm currently relying on Core Storage to encrypt volumes before creating pools on them, and would be glad to get rid of this extra layer...

**Comment by lundman** [comment]
Oh this landed in master and is present in 1.8.1
